README.md
# dcurooms

![TravisCI Status](https://travis-ci.org/theycallmemac/dcurooms.svg?branch=master) [![Maintainability](https://api.codeclimate.com/v1/badges/d41179565d12397e7438/maintainability)](https://codeclimate.com/github/theycallmemac/dcurooms/maintainability) [![GitHub license](https://img.shields.io/github/license/theycallmemac/dcurooms.svg)](https://github.com/theycallmemac/dcurooms/blob/master/LICENSE)

### Description
This is a command line tool I hope to develop overtime. In time it will be able to support the booking of rooms on the DCU Glasnevin Campus. Along with this, it will be able to display room timetables information as well as be able to check the availability of a room in a few different ways.

Currently the dcurooms script can show you which rooms in the Henry Grattan building and the School of Computing are free and booked currently.


### Installation
Clone this repo by running ```git clone https://github.com/theycallmemac/dcurooms.git```.
cd into the clone and run ```python setup.py install``` as root to install.


### Dependencies
 - bs4
 - MechanicalSoup
 - requests


### Usage
Buildings are referenced using their assigned building letters. The School of Computing is given the letter "L" and the Henry Grattan is given the letter "C". Don't ask me why it's like that, it's a very odd naming convention.

To book L125 in the School of Computing on the 7/2/2018 from 16:00 to 17:00, run ```dcurooms -b L125 7/2/2018 1600 1700```.

To book CG04 in the Henry Grattan on the 5/2/2018 from 18:00 to 20:00, run ```dcurooms -b CG04 5/2/2018 1800 2000```.

To check the current availabilty of labs in the DCU School of Computing run ```dcurooms -nL``` or ```dcurooms --now --computing```.

To show the rooms which are currently available in the Henry Grattan building run ```dcurooms -anC``` or ```dcurooms --available --now --grattan```.

To display the information of a specific room in the Henry Grattan building at 16:00 on the Monday of the 4th week of the college year run ```dcurooms -l CG05 4 1 1600``` or ```dcurooms --lookup CG05 4 1 1600```.

To look up the info of all rooms in the School of Computing at 13:00 on the Thursday of the 10th week of the college year run ```dcurooms -lL 10 4 1300``` or ```dcurooms --lookup --computing 10 4 1300```

To lookup the info of all free rooms in the Henry Grattan building at 16:00 on the Monday of the 21st week of the college year run ```dcurooms -laC 21 1 1600``` or ```dcurooms --lookup --available --grattan 21 4 1600```.

For help run ```dcurooms -h``` and for the current version run ```dcurooms --version```.


CHANGELOG.md
# changelog

---

__entry:__ 10-12-2017
__contributor:__ theycallmemac

---

  - README.md
      - created a README markdown file in the intial commit.
      - made some further ammendments to this.

  - LICENSE
      - added GNU General Public License v3.0 to the repo

  - book_lab.py, book_room.py, check_availability.py, booking.sh
      - uploaded previously existing files to new repo.

  - free_now.py, booking.sh
    - created free_now.py and added support for this in the shell script.

  - CHANGELOG.md
    - created a markdown file to document the changes to the code.

---

---

__entry:__ 11-12-2017
__contributor:__ theycallmemac

---

  - README.md
      - changed to include information on setup and how to run the script.

  - setup.py
    - added to install dcurooms script with it's dependencies.
    - changed license info in the script

  - scripts/dcurooms
    - dcurooms script added with the previous 'free_now.py' capabilities.

  - book_lab.py, book_room.py, check_availability.py, booking.sh, free_now.py, booking.sh
    - removed with a view to add their functionaltiies to dcurooms script.

---

---

__entry:__ 13-12-2017
__contributor:__ theycallmemac

---

 
  - scripts/dcurooms
    - now at v0.2.0 follwoing a number of minor changes.
    - changed the "-f"/"--free" option to "-n"/"--now". This is the purposes of clarity in the command name.
    - script now supports a "-f" or "--free" option which specificially displays rooms that are free as opposed to just the status of all rooms. This option checks the info returned from "-n"/"--now".
    - changed license info which was incorrect.
    - rewrote the get_current_time function to calculate the current week as opposed to it being hardcoded into the program.


  - README.md
    - added to long command information
    - gave usage example to check for free rooms rather than all rooms

---

---

__entry:__ 14-12-2017
__contributor:__ theycallmemac

---

  - .gitignore
    - created to ignore .pyc and .txt files
    - also ignores egg-info aswell as build and dist directories

  - scripts/dcurooms
    - created, just import dcurooms.py script

  - scripts/dcurooms.py
    - now at v0.2.1 following a number of fixes changes.
    - import specific tools rather than entire libraries.
    - functions return information ready for concatenation.
    - fixed error for wen the hour returned from get_current_time was greater or equal to 23.

  - ```__init__.py```
    - added one in each directory for the unittests
    - helps navigate around the project
  - tests/test_free_now.py
    - created unittest to check the return values of dcurooms
    - tests for exit code
    - tests if output is empty (indicating free) or a hardcoded string.

---

---

__entry:__ 16-12-2017
__contributor:__ theycallmemac

---

  - scripts/dcurooms.py
    - now at v0.3.0 due to some minor changes in what the tool can do and it's usage.
    - can now look up specific rooms on any given week and timeslot. 
    - can also lookup a whole builing when provided with this info too.
    - the "-f"/"--free" command now replaced with "-a"/"--available" command. This is due to a feature I plan for the next minor update.
    - bugs most definitely persist in this new addition. Fixes on the way soon.

  - README.md
    - added information on the new lookup commands.
    - changed info on previous commands listed.
  
  - setup.py
    - added url to setup script.

  - tests/test_available_now.py
    - tests the -anc options rather that the previous -fnc options.
    - renamed file and changed some names around the test.
    - fixed issue #2

---

---

__entry:__ 19-12-2017
__contributor:__ theycallmemac

---

  - scripts/dcurooms.py
    - now at v0.4.0 because of some rewrites and added functionality
    - lookup option now supports a lookup when available option, given as "-la"/"--lookup --available"
    - search_dictionary function added and repetitive lines of code removed.
    - now gives error message when room entered which either doesn't exist or is not yet supported by the tool, this fixes #4
    - now at v0.4.1 due to some fixes
    - found that issue #7 pertained to more than just what was detailed initially
    - tool no longer returns any IndexErrors instead returns a message as to what you entered wrong.
    - this fixes #7
    - 
  - README.md
    - added information on the new lookup when available command
  
  - tests/test_available_now.py
    - removed unnecessary lines
    - changed information and created txt file based on new option names from v0.3.0

---

---

__entry:__ 20-12-2017
__contributor:__ theycallmemac

---

  - scripts/dcurooms.py
    - now at v0.4.2
    - Refactored script from 214 lines to 149 lines, removing large blocks of repetitive code
    - this refactoring fixes issue #8

---

---

__entry:__ 21-12-2017
__contributor:__ theycallmemac

---

  - scripts/dcurooms.py
    - made dcurooms compatible with python3.x, thus closing issue #6
    - this was done by using MechanicalSoup instead of mechanize as the former supports python2 and python3
    - script also decides between http.cookiejar and cookielib bepending on which major version of python the script was installed with.

  - setup.py
    - included more info on the versions of python that the dcurooms script supports

---

---

__entry:__ 24-12-2017
__contributor:__ theycallmemac

---

  - scripts/dcurooms.py
    - closes #3, milestone 1.0.0 reached
    - supports the booking of rooms in the Henry Grattan Building and labs in the School of Computing.
    - requires your gmail username and password to send lab booking on behalf of your email.
    - supports both python2 and python3

---
---

__entry:__ 25-12-2017
__contributor:__ theycallmemac

---

  - scripts/dcurooms.py
    - fixed some errors persisting within the booking code, incorrect parameters etc

  - tests/test_available_now.py
    - made changes to test, giving exit status of 0 rather than 1.

scripts/index.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# import libraries
import sys
import datetime
import smtplib
from optparse import OptionParser
from mechanicalsoup import StatefulBrowser
from requests import get
from bs4 import BeautifulSoup
sys.path.append(".")
from scripts import controls
if sys.version_info[0] < 3:
    from cookielib import LWPCookieJar
else:
    from http.cookiejar import LWPCookieJar

__author__ = "theycallmemac"
__version__ = '1.0.0'
__copyright__ = 'Copyright (c) 2017 theycallmemac'
__license__ = 'GPL-3.0'


def setup_options():
    parser = OptionParser(
        description='Displays info and books room around DCU.',
        prog='dcurooms', version='%prog ' + __version__,
        usage='%prog [option]')
    parser.add_option(
        "-l", "--lookup", action="store_true",
        help="returns information given a specific room, week, day and hour")
    parser.add_option(
        "-b", "--book", action="store_true",
        help="books a room by providing the room, D/M/YYYY, start, and end")
    parser.add_option(
        "-a", "--available", action="store_true",
        help="returns only the rooms/labs that are free in a building")
    parser.add_option(
        "-n", "--now", action="store_true",
        help="show the status of each room/lab as it is currently")
    # building options which can be passed
    parser.add_option(
        "-L", "--computing", action="store_true",
        help="displays the status of the labs in the School of Computing")
    parser.add_option(
        "-C", "--grattan", action="store_true",
        help="displays the status of rooms in the Henry Grattan building")
    return parser

def get_data():
    times, c, g, details = {'0800': '1', '0830': '2', '0900': '3', '0930': '4', '1000': '5', '1030': '6', '1100': '7', '1130': '8', '1200': '9', '1230': '10', '1300': '11', '1330': '12', '1400': '13','1430': '14', '1500': '15', '1530': '16', '1600': '17', '1630': '18', '1700': '19', '1730': '20', '1800': '21', '1830': '22', '1900': '23', '1930': '24', '2000': '25', '2030': '26', '2100': '27', '2130': '28', '2200': '29', '2230': '30'}, ['LG25', 'LG26', 'LG27', 'L101', 'L114', 'L125', 'L128'], ['CG01', 'CG02', 'CG03', 'CG04', 'CG05', 'CG06', 'CG11', 'CG12', 'CG20', 'CG68', 'C166'], sys.argv[2:]
    return times, c, g, details

def get_lst(c, g, options):
    if options.computing: lst = c
    elif options.grattan: lst = g
    return lst

def opt_is_room(options, rooms, details, times):
    c, g = rooms[0:2]
    if options.computing: lst = c
    elif options.grattan: lst = g
    else:
        controls.lookup_room_control(g, c, details, times)

def booking_option(rooms, details):
    c, g = rooms[0:2]
    print("Booking requires more arguments. See the help for details.") if len(details) < 4 else controls.booking_control(c, g, details)
    sys.exit()
    
def lookup_option(options, rooms, details, times):
    opt_is_room(options, rooms, details, times)
    lst = get_lst(rooms[0], rooms[1], options)
    controls.lookup_building_control(options, lst, details, times)
    
def now_option(options, rooms, times):
    c, g = rooms[0:2]
    lst = get_lst(c, g, options)
    controls.available_now_control(options, lst, times)
    
def required():
    parser = setup_options()
    (options, arguments) = parser.parse_args()
    times, c, g, details = get_data()
    return parser, (options, arguments), (c, g), (times, details)

def requesting_booking(options, rooms, details):
    if options.book and len(details) < 5:
        booking_option(rooms, details)
        
def displaying_info(options, rooms, details, times):
    if options.lookup:
        lookup_option(options, rooms, details, times)
    elif options.now:
        now_option(options, rooms, times)

def flow(options, rooms, info):
    times, details = info[0:2]
    requesting_booking(options, rooms, details)
    displaying_info(options, rooms, details, times)

def main():
    parser, (options, arguments), rooms, info = required()
    flow(options, rooms, info)
    parser.print_help()
if __name__ == '__main__':
    main()

